### PR TITLE
Introduction to Dynamic Memory Allocation

### PR DESCRIPTION
**Pull request by jerrettl**

### CODE REVIEW

**Comment by jerrettl** [comment]
This introduction to dynamic memory allocation is based on the slides that Idel and I made in Fall 2020. I have added some new information here and there, but the scope of this section is just for a basic introduction to DMA, without getting too deep into the code yet. I may add a note that this is just for getting your feet wet into DMA. I intend to break down further discussions on dynamic memory into another page, like for basic use of `malloc()` and `free()` as well as pointer safety.

I am also considering fleshing out the `calloc()` and `realloc()` sections since while they aren't covered in some Intro to C classes, usage varies depending on the course and instructor. A formal explanation for these can be very helpful for students.

File: src/intro-to-c/dma/intro-to-dma.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+This still causes an infinite loop, but at least you won't be hogging all the memory anymore.
+
+---
+
+## `calloc()`: The Over-Accomplishing Sibling
+
+`calloc()` is like the sibling of `malloc()` that always tries to do more. They are both related, but have differences on what they do. `calloc()` is short for "**c**ontiguous **alloc**ation," and it will allocate memory just like `malloc()`, but also **initialize** the memory to 0, so you have no garbage values right off the bat. In this way, `calloc()` can be useful for making, for example, arrays that would have been initialized anyway.
+
+`calloc()` takes a size and a quantity of that size. For example:
```
Comment (ZThatWhiteGuyZ) [inline src/intro-to-c/dma/intro-to-dma.md]: I would switch the order of the wording, just so it lines up with the syntax. Quantity then size.

File: src/intro-to-c/dma/intro-to-dma.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    malloc(sizeof(int));
+  }
+
+  return 0;
+}
+```
+
+Nooooo! This is all wrong.
+
+This program is continuing to allocate memory *without* returning it to other programs. It is *such* a hog, leaking memory all over the floor. Don't be a menace to your computer.
```
Comment (ZThatWhiteGuyZ) [inline src/intro-to-c/dma/intro-to-dma.md:135]: Good callback. Much appreciated.

**Comment by ZThatWhiteGuyZ** [comment]
Overall looks good. I agree with some more about `calloc()` and `realloc()`, at least because they're interesting. Nothing wrong from learning from your supplementary material.

File: src/intro-to-c/dma/intro-to-dma.md (hunk 1 above)
Comment (jerrettl) [reply src/intro-to-c/dma/intro-to-dma.md]: Agreed.

**Comment by jerrettl** [comment]
The basic intro to DMA is done. Further lessons will break down what's actually happening when you dynamically allocate memory.

**Pull request merged**